train.c

```diff
@@ -0,0 +1,116 @@
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <float.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include "agents/temporal_difference.h"
+#include "game.h"
+
+#define NUM_ITERATION 10000
+#define EPSILON_START 0.5
+#define EPSILON_END 0.001
+static double epsilon = EPSILON_START;
+static double decay_factor;
+static unsigned int N_STATES = 1;
+static td_agent_t agent[2];
+#define RAND_UNIFORM ((float) rand() / (float) RAND_MAX)
+
+static void init_training()
+{
+    srand(time(NULL));
+    CALC_STATE_NUM(N_STATES);
+    init_td_agent(&agent[0], N_STATES, 'O');
+    init_td_agent(&agent[1], N_STATES, 'X');
+    decay_factor =
+        pow((EPSILON_END / EPSILON_START), (1.0 / (NUM_ITERATION * N_GRIDS)));
+}
+
+static int *get_available_moves(char *table, int *ret_size)
+{
+    int move_cnt = 0;
+    int *available_moves = malloc(N_GRIDS * sizeof(int));
+    if (!available_moves) {
+        perror("Failed to allocate memory");
+        exit(1);
+    }
+    for_each_empty_grid(i, table)
+    {
+        available_moves[move_cnt++] = i;
+    }
+    *ret_size = move_cnt;
+    return available_moves;
+}
+
+static int get_action_epsilon_greedy(char *table, td_agent_t *agent)
+{
+    int move_cnt = 0;
+    int *available_moves = get_available_moves(table, &move_cnt);
```

Comment:
> I'm sorry that I didn't think clearly. We can't obtain the random space straightly. We need to complete it in three steps:
> 
> 1. Get the number of spaces n by `for_each_empty_grid`.
> 2. Get a random count i between [0, n).
> 3. Get the board index of the ith space by `for_each_empty_grid`.
> 
> This method has the same time complexity as the original one but is not very elegant.

However, I still prefer my proposed method to get a random empty cell because it's in-place.

Reply:
I understand your thought process. However, your method requires a two-pass loop, and I'm not sure which way is better here.

Reply:
If someday we wanna present the board as a bit array instead of a char array to save memory usage, the `get_available_moves` may be the bottleneck because it always outputs an int array bigger than the board.

Reply:
I got it. It indeed makes sense. I thought you could submit these changes with another PR since these changes provide a different aspect to improve this project, and it is worth addressing your proposal in detail.